File: src/main/java/view/RootView.java
package view;

import domain.*;
import org.jetbrains.annotations.NotNull;
import org.jline.terminal.Terminal;

import java.io.IOException;
import java.io.InputStream;
import java.io.PrintStream;

public final class RootView {
    private @NotNull Terminal      mTerminal;
    private @NotNull TileBuffer    mBuffer;
    private @NotNull Flash         mFlash;
    private @NotNull GameView      mGameView;
    private @NotNull PickPawnInput mPickPawnInput;

    public RootView(@NotNull Terminal terminal) {
        mTerminal = terminal;
        mFlash    = new Flash(terminal);
        mBuffer   = new TileBuffer(terminal);
        mGameView = new GameView(mBuffer);
        mPickPawnInput = new PickPawnInput(terminal);
    }

    // commands
    public Integer pickPawnId() throws IOException, InterruptedException   {
        mPickPawnInput.call();
        return mPickPawnInput.getSelectedPawnIndex();
    }

    // render
    public void renderGame(@NotNull Game game) {
        mGameView.render(game, mPickPawnInput.getSelectedPawnIndex());
    }

    public void renderError(@NotNull Error error) {
        mFlash.render(error);
    }

    // draw
    public void clear() {
        mTerminal.writer().print("\033[H\033[2J");
        mTerminal.writer().flush();
    }

    public void draw() {
        mBuffer.draw();
        mFlash.draw();
        mTerminal.flush();
    }
}

File: Sources/Onitama/view/RootView.swift
final class RootView {
  private let mTerminal: Terminal
  private let mBuffer: TileBuffer
  private let mGameView: GameView
  private let mFlash: Flash
  private let mPickPawnInput: PickPawnInput

  // -- lifetime --
  init(terminal: Terminal) {
    mTerminal = terminal
    mBuffer = TileBuffer(terminal: terminal)
    mGameView = GameView(buffer: mBuffer)
    mFlash = Flash(terminal: terminal)
    mPickPawnInput = PickPawnInput(terminal: terminal)
  }

  // -- commands --
  func pickPawnId() -> Int? {
    mPickPawnInput.call()
    return mPickPawnInput.confirmedPawnIndex
  }

  // -- commands/render
  func render(game: Game) {
    mGameView.render(game: game, selectedPawnIndex: mPickPawnInput.selectedPawnIndex)
  }

  func render(error: PendingMove.Error) {
    mFlash.render(error: error)
  }

  // -- commands/draw
  func clear() {
    mTerminal.write("\u{001B}[H\u{001B}[2J")
  }

  func draw() {
    mBuffer.draw()
    mFlash.draw()
  }
}
